[New Tutorial] Designing Scalable and Versatile Storage Solutions on Amazon EKS with the Amazon EFS CSI (#748)
build-on-aws/content: tutorials/eks-with-efs-add-on/index.md

--role-name $ROLE_NAME \
--role-only \
--attach-policy-arn arn:aws:iam::aws:policy/service-role/AmazonEFSCSIDriverPolicy \
--approve
```

It takes a few minutes for the operation to complete. The expected output should look like this:

```text
2023-08-16 15:27:19 [ℹ]  4 existing iamserviceaccount(s) (cert-manager/cert-manager,default/external-dns,kube-system/aws-load-balancer-controller,kube-system/efs-csi-controller-sa) will be excluded
2023-08-16 15:27:19 [ℹ]  1 iamserviceaccount (kube-system/efs-csi-controller-sa) was excluded (based on the include/exclude rules)
2023-08-16 15:27:19 [!]  serviceaccounts in Kubernetes will not be created or modified, since the option --role-only is used
2023-08-16 15:27:19 [ℹ]  no tasks
```

Run the following command to add the Kubernetes service account name to the trust policy for the IAM role:

```bash
export TRUST_POLICY=$(aws iam get-role --role-name $ROLE_NAME --query 'Role.AssumeRolePolicyDocument' | \
sed -e 's/efs-csi-controller-sa/efs-csi-*/' -e 's/StringEquals/StringLike/')
```

Update the IAM role:

```bash
aws iam update-assume-role-policy --role-name $ROLE_NAME --policy-document "$TRUST_POLICY" --region $CLUSTER_REGION
```

> ## Step 3: Verify or Install the EFS CSI Driver add-on

Reviewer: ```suggestion
## Step 3: Verify or Install the EFS CSI Driver Add-on
```

    --group-name MyEfsSecurityGroup \
    --description "My EFS security group" \
    --vpc-id $CLUSTER_VPC \
    --region $CLUSTER_REGION \
    --output text)
```

3. Create an inbound rule that allows inbound NFS traffic from the CIDR for your cluster's VPC.

```bash
aws ec2 authorize-security-group-ingress \
    --group-id $SECURITY_GROUP_ID \
    --protocol tcp \
    --port 2049 \
    --cidr $CIDR_RANGE \
    --region $CLUSTER_REGION
```

The expected output should look like this:

```text
{
    "Return": true,
    "SecurityGroupRules": [
        {
            "SecurityGroupRuleId": "sgr-0106e3f85cd3b82d9",
            "GroupId": "sg-0a0a11596fe0ae56f",
            "GroupOwnerId": "000866617021",
            "IsEgress": false,
            "IpProtocol": "tcp",
            "FromPort": 2049,
            "ToPort": 2049,
            "CidrIpv4": "192.168.0.0/16"
        }
    ]
}
```

Now, we’ll create an Amazon EFS File System for our Amazon EKS cluster.

```bash
export FILE_SYSTEM_ID=$(aws efs create-file-system \
--region $CLUSTER_REGION \
--performance-mode generalPurpose \
--query 'FileSystemId' \
--output text)
```

## Step 5: Configure Mount Targets for the EFS File System

> EFS only allows one mount target to be created in each Availability Zone, so you’ll need to place the mount target on the appropriate subnet. If your worker nodes are on a private subnet, you should create the mount target on that subnet. We will now create mount targets for the EFS File System, The mount target is an IP address on a VPC subnet that accepts NFS traffic. The Kubernetes nodes will open NFS connections with the IP address of the mount target.

Reviewer: ```suggestion
EFS only allows one mount target to be created in each Availability Zone, so you’ll need to place the mount target on the appropriate subnet. If your worker nodes are on a private subnet, you should create the mount target on that subnet. We will now create mount targets for the EFS File System, the mount target is an IP address on a VPC subnet that accepts NFS traffic. The Kubernetes nodes will open NFS connections with the IP address of the mount target.
```

|  region-codeb    |  192.168.0.0/19    |  subnet-EXAMPLE123c7c5182  |
|  region-codea    |  192.168.160.0/19  |  subnet-EXAMPLE0416ce588p  |
+------------------+--------------------+----------------------------+
```

3. **Add Mount Targets** for the Subnets Hosting Your Nodes: Run the following command to create the mount target, specifying each subnet. For example, if the cluster has a node with an IP address of 192.168.56.0, and this address falls within the CidrBlock of the subnet ID subnet-EXAMPLEe2ba886490, create a mount target for this specific subnet. Repeat this process for each subnet in every Availability Zone where you have a node, using the appropriate subnet ID.

```bash
aws efs create-mount-target \
    --file-system-id $FILE_SYSTEM_ID \
    --subnet-id subnet-EXAMPLEe2ba886490 \
    --security-groups $SECURITY_GROUP_ID \
    --region $CLUSTER_REGION
```

>Note: EFS only allows 1 mount target to be created in one Availability Zone, irrespective of the subnets in the Availability Zone. If you are using an EKS cluster with Worker Nodes in Private Subnets, it would be recommended to create the mount targets for the same subnets.

> ## Step 6: Setup a Storage Class for the Sample Application

Reviewer: ```suggestion
## Step 6: Set Up a Storage Class for the Sample Application
```

In Kubernetes, there are two ways to provision storage for container applications: Static Provisioning and Dynamic Provisioning. In Static Provisioning, a cluster administrator manually creates Persistent Volumes (PVs) that specify the available storage for the cluster's users. These PVs are part of the Kubernetes API and can be easily used. On the other hand, Dynamic Provisioning automates the creation of PVs. Kubernetes uses Storage Classes to generate PVs automatically when a pod requests storage through PersistentVolumeClaims. This method simplifies the provisioning process and adjusts to the specific needs of the application. To learn more, refer [PersistentVolumes](https://kubernetes.io/docs/concepts/storage/persistent-volumes/) in Kubernetes documentation. For our lab, we will use the “Dynamic Provisioning Method” method.

> 

Reviewer: ```suggestion
```

1. First, download the Storage Class manifest:

```bash
curl -O https://raw.githubusercontent.com/kubernetes-sigs/aws-efs-csi-driver/master/examples/kubernetes/dynamic_provisioning/specs/storageclass.yaml
```

2. Verify that the ```$FILE_SYSTEM_ID``` environment variable is configured:

```bash
echo $FILE_SYSTEM_ID
```

3. Update the FileSystem identifier in the storageclass.yaml manifest:

```bash
sed -i 's/fs-92107410/$FILE_SYSTEM_ID/g' storageclass.yaml
```

4. **Optionally**, if you're running this on macOS, run the following command to change the file system:

```bash
sed -i '' 's/fs-92107410/'"$FILE_SYSTEM_ID"'/g' storageclass.yaml
```

> 5. Deploy the StorageClass: 

Reviewer: Should this be one word, or Storage Class? You have it as "Storage Class" a few lines above.

Author: Have changed it, where there is one word, its majorly for filename purposes or kubectl commands

4. Verify that the data is being written to the /data/out location within the shared EFS volume. Use the following command to verify:

```bash
kubectl exec efs-app -- bash -c "cat data/out"
```

The expected output should look like this:

```text
[...]
Fri Aug 4 09:13:40 UTC 2023
Fri Aug 4 09:13:45 UTC 2023
Fri Aug 4 09:13:50 UTC 2023
Fri Aug 4 09:13:55 UTC 2023
[...]
```

>Optionally, terminate the node hosting your pod and await pod rescheduling. Alternatively, you may delete the pod and redeploy it. Repeat the previous step once more, ensuring the output contains the prior output.

## Clean Up

To avoid incurring future charges, you should delete the resources created during this tutorial.

```text
# Delete the pod
kubectl delete -f pod.yaml

# Delete the Storage Class
kubectl delete -f storageclass.yaml

# Delete the EFS CSI add-on
eksctl delete addon --cluster $CLUSTER_NAME --name $ADD_ON

# Delete the IAM Role
aws iam delete-role --role-name AmazonEKS_EFS_CSI_DriverRole
```

## Conclusion

> With the completion of this tutorial, you have successfully configured Amazon EFS for persistent storage for your EKS-based container workloads. The sample pod, leveraging the CentOS image, has been configured to capture the current date and store it in the ```/data/out``` directory on the EFS shared volume. To align with best practices, we recommend running your container workloads on private subnets, exposing only ingress controllers in the public subnets. Furthermore, make sure that EFS mount targets are established in all availability zones where your EKS cluster resides; otherwise, you may encounter a 'Failed to Resolve' error. To continue your journey, you're now ready to store your stateful workloads like batch processes or machine learning training data to your EFS volume. 

Reviewer: Can you add 2-3 links to other pieces that the curious reader may want to check out next?

Author: Added links to the above section. some of them are preview links for the blogs that are supposed to go live with this blog as well. Hopefully, the links can be changed at your end to the original public link

> ---

Reviewer 2: Hello!

What a well-written blog! It's fluid and the explanations are pithy, which make it easy to follow. 

My edits:

- Corrected titles to match AP Style. 
- I picked up on "Amazon EKS Driver Add-on" and "Amazon EKS Driver Add-On". Amazon has it listed as "Amazon EKS add-ons"--without capitalization on the "add-on". Please uniform these. 
- Quite a few sections start with "in this section". I suggested a rewrite to break it up and make it more conversational. 
- 
Please see below for more notes and suggestions. 

Overall, well done.  👍

title: "Designing Scalable and Versatile Storage Solutions on Amazon EKS with the Amazon EFS CSI"
description: "How to configure persistent shared storage for container workloads on Amazon EKS using the Amazon Elastic File System (EFS) CSI Driver Add-On."

> tags:

Author: done

    - eks-cluster-setup
    - eks
    - kubernetes
    - tutorials
    - aws
    - storage
showInHomeFeed: true
waves:
  - modern-apps
spaces:
  - kubernetes
  - modern-apps
authorGithubAlias: ashishkamble4
authorName: Ashish Kamble
date: 2023-09-30
---

Managing storage solutions for containerized applications requires careful planning and execution. Kubernetes workloads such as content management systems and video transcoding may benefit from using Amazon Elastic File System (EFS). EFS is designed to provide serverless, fully elastic file storage that lets you share file data without provisioning or managing storage capacity and performance. EFS is ideal for applications that need shared storage across multiple nodes or even across different availability zones. In contrast, Amazon Elastic Block Store (EBS) requires configuring volumes that are limited by size and region. The Amazon Elastic File System (EFS) CSI Driver Add-On exposes EFS file systems to your workloads, handling the complexities of this versatile and scalable storage solution.

> In this tutorial, you will set up the Amazon EFS CSI Driver on your Amazon EKS cluster and configure persistent shared storage for container workloads. More specifically, you will install the EFS CSI Driver as an Amazon EKS Add-on. Amazon EKS add-ons automate installation and management of a curated set of add-ons for EKS clusters. You'll configure the EFS CSI Driver to handle both lightweight applications, akin to microservices, and more substantial systems, comparable to databases or user authentication systems, achieving seamless storage management. 

Author: done

* Install the latest version of [kubectl](https://kubernetes.io/docs/tasks/tools/#kubectl). To check your version, run: `kubectl version --short`.
* Install the latest version of [eksctl](https://eksctl.io/introduction/#installation). To check your version, run: `eksctl info`.

## Overview

This tutorial is the second installment in a series on optimizing stateful applications in Amazon EKS, focusing on the Amazon EFS CSI Driver for managing complex containerized applications that require persistent and shared storage content across availability zones. This tutorial not only guides you through the configuration of the EFS CSI Driver in the "kube-system" namespace but also delves into the intricacies of storage management within the cluster. It covers the following components:

* **Authentication**: Leverage the pre-configured IAM Role for the Amazon EFS CSI Driver, integrated with the OpenID Connect (OIDC) endpoint, to ensure secure communication between Kubernetes pods and AWS services.
* **EFS CSI Driver Setup**: Deploy the Amazon EFS CSI Driver within the Amazon EKS cluster, focusing on Custom Resource Definitions (CRDs) and the installation of the driver itself.

> * **Sample Application Deployment**: Build and deploy a stateful application that writes the current date to a shared EFS volume. Define routing rules and annotations for a Persistent Volume Claim (PVC) based on the CentOS image. Utilize custom annotations for the PVC, specifically the 'accessMode' and 'storageClassName', to instruct the EFS CSI Driver on how to handle storage requests. For Dynamic Provisioning, use the 'storageClassName' annotation to automate the creation of Persistent Volumes (PVs)

Author: done

* **EFS Access Points and Security**: Explore how the Amazon EFS CSI Driver leverages Amazon EFS access points as application-specific entryways. Understand the role of port 2049 for NFS traffic and how to configure security groups to allow or restrict access based on CIDR ranges. These access points enforce both user identity and root directory, offering flexibility to further refine access based on specific subnets. By fine-tuning the security group to permit traffic on this port, you empower the instances or pods within your designated CIDR range to interact with the shared file system. This granular control not only ensures the accessibility of the EFS by relevant cluster resources but also allows for further access restrictions based on specific subnets.

> * **Mount Targets**: Delve into the creation of mount targets for the EFS Filesystem, which serve as crucial connection points within the VPC for EC2 instances. It's essential to note that EFS only allows one mount target to be created in each Availability Zone, regardless of the subnets within that zone. If you are using an EKS cluster with nodes in private subnets, we recommend creating the Mount targets for those specific subnets.

Author: done

2023-08-16 15:27:19 [ℹ]  no tasks
```

Run the following command to add the Kubernetes service account name to the trust policy for the IAM role:

```bash
export TRUST_POLICY=$(aws iam get-role --role-name $ROLE_NAME --query 'Role.AssumeRolePolicyDocument' | \
sed -e 's/efs-csi-controller-sa/efs-csi-*/' -e 's/StringEquals/StringLike/')
```

Update the IAM role:

```bash
aws iam update-assume-role-policy --role-name $ROLE_NAME --policy-document "$TRUST_POLICY" --region $CLUSTER_REGION
```

## Step 3: Verify or Install the EFS CSI Driver Add-On

In this section, we'll verify that the EFS CSI Driver managed add-on is properly installed and active on your Amazon EKS cluster. The EFS CSI Driver is crucial for enabling Amazon EFS to work seamlessly with Kubernetes, allowing you to mount EFS file systems as persistent volumes for your batch workloads.

Check whether the EFS CSI driver add-on is installed on your cluster:

```bash
eksctl get addon --cluster ${CLUSTER_NAME} --region ${CLUSTER_REGION} | grep efs
```

The expected output should look like this:

```text
aws-efs-csi-driver      v1.5.8-eksbuild.1       ACTIVE  0
```

> **Optionally**, if the EFS CSI Driver Add-On is **not** installed on your cluster, or you receive an error, the following commands show you the steps to install it. Note that you must have an [OpenID Connect (OIDC) endpoint](https://docs.aws.amazon.com/eks/latest/userguide/enable-iam-roles-for-service-accounts.html?sc_channel=el&sc_campaign=appswave&sc_content=eks-cluster-load-balancer-ipv4&sc_geo=mult&sc_country=mult&sc_outcome=acq) associated with your cluster before you run these commands.

Reviewer 2: Please uniform "add-on".

Author: done

export SECURITY_GROUP_ID=$(aws ec2 create-security-group \
    --group-name MyEfsSecurityGroup \
    --description "My EFS security group" \
    --vpc-id $CLUSTER_VPC \
    --region $CLUSTER_REGION \
    --output text)
```

3. Create an inbound rule that allows inbound NFS traffic from the CIDR for your cluster's VPC.

```bash
aws ec2 authorize-security-group-ingress \
    --group-id $SECURITY_GROUP_ID \
    --protocol tcp \
    --port 2049 \
    --cidr $CIDR_RANGE \
    --region $CLUSTER_REGION
```

The expected output should look like this:

```text
{
    "Return": true,
    "SecurityGroupRules": [
        {
            "SecurityGroupRuleId": "sgr-0106e3f85cd3b82d9",
            "GroupId": "sg-0a0a11596fe0ae56f",
            "GroupOwnerId": "000866617021",
            "IsEgress": false,
            "IpProtocol": "tcp",
            "FromPort": 2049,
            "ToPort": 2049,
            "CidrIpv4": "192.168.0.0/16"
        }
    ]
}
```

Now, we’ll create an Amazon EFS file system for our Amazon EKS cluster.

```bash
export FILE_SYSTEM_ID=$(aws efs create-file-system \
--region $CLUSTER_REGION \
--performance-mode generalPurpose \
--query 'FileSystemId' \
--output text)
```

> ## Step 5: Configure Mount Targets for the EFS Filesystem

Author: done

> EFS only allows one mount target to be created in each Availability Zone, so you’ll need to place the mount target on the appropriate subnet. If your worker nodes are on a private subnet, you should create the mount target on that subnet. In this section, you will create mount targets for the EFS filesystem. The mount target is an IP address on a VPC subnet that accepts NFS traffic. The Kubernetes nodes will open NFS connections with the IP address of the mount target.

Author: done

+------------------+--------------------+----------------------------+
|  region-codec    |  192.168.128.0/19  |  subnet-EXAMPLE6e421a0e97  |
|  region-codeb    |  192.168.96.0/19   |  subnet-EXAMPLEd0503db0ec  |
|  region-codec    |  192.168.32.0/19   |  subnet-EXAMPLEe2ba886490  |
|  region-codeb    |  192.168.0.0/19    |  subnet-EXAMPLE123c7c5182  |
|  region-codea    |  192.168.160.0/19  |  subnet-EXAMPLE0416ce588p  |
+------------------+--------------------+----------------------------+
```

3. **Add Mount Targets** for the Subnets Hosting Your Nodes: Run the following command to create the mount target, specifying each subnet. For example, if the cluster has a node with an IP address of 192.168.56.0, and this address falls within the CidrBlock of the subnet ID subnet-EXAMPLEe2ba886490, create a mount target for this specific subnet. Repeat this process for each subnet in every Availability Zone where you have a node, using the appropriate subnet ID.

```bash
aws efs create-mount-target \
    --file-system-id $FILE_SYSTEM_ID \
    --subnet-id subnet-EXAMPLEe2ba886490 \
    --security-groups $SECURITY_GROUP_ID \
    --region $CLUSTER_REGION
```

> >Note: EFS only allows 1 mount target to be created in one Availability Zone, irrespective of the subnets in the Availability Zone. If you are using an EKS cluster with Worker Nodes in Private Subnets, it would be recommended to create the Mount targets for the same subnets.

Author: done

## Step 6: Setup a Storage Class for the Sample Application

> In Kubernetes, there are two ways to provision storage for container applications: Static Provisioning and Dynamic Provisioning. In Static Provisioning, a cluster administrator manually creates Persistent Volumes (PVs) that specify the available storage for the cluster's users. These PVs are part of the Kubernetes API and can be easily used. On the other hand, Dynamic Provisioning automates the creation of PVs. Kubernetes uses Storage Classes to generate PVs automatically when a Pod requests storage through PersistentVolumeClaims. This method simplifies the provisioning process and adjusts to the specific needs of the application. To learn more, refer [PersistentVolumes](https://kubernetes.io/docs/concepts/storage/persistent-volumes/) in Kubernetes documentation. In this section, we will use the “Dynamic Provisioning Method” method.

Author: done

1. First, download the Storage Class manifest:

```bash
curl -O https://raw.githubusercontent.com/kubernetes-sigs/aws-efs-csi-driver/master/examples/kubernetes/dynamic_provisioning/specs/storageclass.yaml
```

2. Verify that the ```$FILE_SYSTEM_ID``` environment variable is configured:

```bash
echo $FILE_SYSTEM_ID
```

3. Update the FileSystem identifier in the storageclass.yaml manifest:

```bash
sed -i 's/fs-92107410/$FILE_SYSTEM_ID/g' storageclass.yaml
```

4. **Optionally**, if you're running this on macOS, run the following command to change the file system:

```bash
sed -i '' 's/fs-92107410/'"$FILE_SYSTEM_ID"'/g' storageclass.yaml
```

5. Deploy the StorageClass: 

```bash 
kubectl apply -f storageclass.yaml
```

The expected output should look like this:

```text
storageclass.storage.k8s.io/efs-sc created
```

## Step 7: Deploy a Sample Application

> In this section, we will deploy a sample application that writes the current date to a shared location on the EFS Shared Volume. The deployment process begins by downloading a manifest file containing the specifications for a Persistent Volume Claim (PVC) and a Pod based on the CentOS image.

Author: done

```bash
kubectl get pods -o wide
```

The expected output should look like this:

```text
NAME          READY   STATUS    RESTARTS   AGE   IP               NODE                                             NOMINATED NODE   READINESS GATES
efs-app       1/1     Running   0          10m   192.168.78.156   ip-192-168-73-191.region-code.compute.internal   <none>           <none>
```

>Note: It will take a couple minutes for the pod to transition to the "Running" state. If the pod remains in a "ContainerCreating" state, make sure that you have included a mount target for the subnet where your node is located (as shown in step two). Without this step, the pod will remain stuck in the "ContainerCreating" state.

4. Verify that the data is being written to the /data/out location within the shared EFS volume. Use the following command to verify:

```bash
kubectl exec efs-app -- bash -c "cat data/out"
```

The expected output should look like this:

```text
[...]
Fri Aug 4 09:13:40 UTC 2023
Fri Aug 4 09:13:45 UTC 2023
Fri Aug 4 09:13:50 UTC 2023
Fri Aug 4 09:13:55 UTC 2023
[...]
```

> >Optionally, terminate the node hosting your Pod and await Pod rescheduling. Alternatively, you may delete the Pod and redeploy it. Repeat the previous step once more, ensuring the output contains the prior output.

Author: Done

## Conclusion

> With the completion of this tutorial, you have successfully configured Amazon EFS for persistent storage for your EKS-based container workloads. The sample Pod, leveraging the CentOS image, has been configured to capture the current date and store it in the ```/data/out``` directory on the EFS shared volume. To align with best practices, we recommend running your container workloads on private subnets, exposing only ingress controllers in the public subnets. Furthermore, make sure that EFS mount targets are established in all availability zones where your EKS cluster resides; otherwise, you may encounter a 'Failed to resolve' error. To continue your journey, you're now ready to store your stateful workloads like batch processes or machine learning training data to your EFS volume. 

Author: done

> ## Clean Up

Author: Done

To avoid incurring future charges, you should delete the resources created during this tutorial.

```text
# Delete the Pod
kubectl delete -f pod.yaml

# Delete the Storage Class
kubectl delete -f storageclass.yaml

> # Delete the EFS CSI Add-On

Reviewer 2: Please uniform "add-on".

Author: done
